### ISSUE TITLE
Not detecting tsconfig.json (Sublime Text 3)

### ISSUE THREAD

**Issue by robertdp** [Microsoft/TypeScript-Sublime-Plugin#213]
I have the file `./tsconfig.json` in the root folder of the project, with `"modules": "commonjs"` set correctly.

When editing any file 2 or more folders deep (e.g. `./src/journal/module.ts`), TypeScript reports an error for every import and export statement:

> Cannot compile modules unless the '--module' flag is provided

This error doesn't show up for files under `./src`, and there are no module/import errors when building the project.

Also mentioned in #207.

**Comment by LPGhatguy** [comment]
This is also reported in issue #180, which is marked as resolved. I'm not sure whether there has been a plugin release with the fix.

**Comment by zhengbli** [comment]
The problem is that currently the server seems only search for two levels of directory, therefore when it goes deeper than that it cannot find the config file. Working on a fix.

@LPGhatguy The issue in #180 is different though. That one is caused by changes in the tsserver.js file that leads to missing config file even in the same directory.

**Comment by corruptmem** [comment]
I'm also experiencing the same problem. I'll try to see if I can get more information about why.

**Comment by LPGhatguy** [comment]
I'm actually experiencing this issue too, and my path to my entrypoint is `one/two/three/scripts/main.ts`, with my deepest paths being `one/two/three/scripts/Components/Core/component.ts`. I get intellisense just fine, and I only get import/export errors sometimes, and I can't seem to reproduce *when* that happens.

**Comment by zhengbli** [comment]
The reason is that currently we don't have a file watcher for tsconfig.json, therefore you have to restart sublime to apply your changes

**Comment by LPGhatguy** [comment]
I just opened my project fresh today (I just got to work) and I'm getting the same errors on one of my files -- I haven't changed `tsconfig.json` in a week or so and I just opened Sublime Text fresh.

**Comment by zhengbli** [comment]
Can you paste the content of your tsconfig.json file? Or do you see any error if you directly run ``tsc`` from the command line in the folder?

**Comment by LPGhatguy** [comment]
I'm just using this file. I don't specify any files; is that a problem?

```json
{
	"compilerOptions": {
		"module": "commonjs",
		"detectGlobals": false
	}
}
```

**Comment by zhengbli** [comment]
@LPGhatguy When I run tsc directly from the commandline, I got the error: "error TS5023: Unknown compiler option 'detectGlobals'." It seems your tsconfig file is not valid

**Comment by LPGhatguy** [comment]
That could be the issue; that was an option I was passing through to a gulp TypeScript transpilation plugin. I'll let you know if I see the issue again with that config.

**Comment by LPGhatguy** [comment]
This issue is back for me! My project file is just

```json
{
	"compilerOptions": {
		"module": "commonjs"
	},
	"files": [
		"typings/deps.d.ts"
	]
}
```

There are no compiler errors with `tsc` and I've reloaded Sublime Text 3 completely a few times to make sure it wouldn't be the issue. I went through my activity monitor to see if there were any lingering io.js 3.1 processes and there were not.

**Comment by thr0w** [comment]
I'm experiencing the same problem, but sometimes works sometimes not works.
I think this bug is in tsserver because I edited `typescript/commands/build.py` with
```python
    print(project_info)
```
and body was empty
```js
{'seq': 0, 
  'command': 'projectInfo', 
  'type': 'response', 
  'body': {}, 
  'request_seq': 7, 
  'success': True
}
```

I dont know how tsserver works, someone can help me?

**Comment by LPGhatguy** [comment]
I hit the issue again after refactoring my typing files to be inputs to the compiler instead of ```/// reference` directives

**Comment by SummerWish** [comment]
I have the same issue here.

If my `tsconfig.json` doesn't contain `files`, or contain `files` and the current file name is in the list, the `tsconfig.json` path is resolved correctly in `typescript/commands/build.py`. Otherwise (contain `files` array but the current file name is not in the array), the `body` is empty:

```json
{'success': True, 'request_seq': 22, 'type': 'response', 'seq': 0, 'command': 'projectInfo', 'body': {}}
```

**Comment by tomachristian** [comment]
Do an

```bash
npm install typescript
```

in the root folder, this will install the latest TypeScript compiler and should work.

**Comment by NeWbY100** [comment]
I had the same problem; I would get the error no matter where I would place a .ts file.
After analyzing what happened, I realized that the TypeScript-Sublime-Plugin uses it's own internal  tsserver.js installed with it. The problem was that it wasn't the same version as the typescript that I had installed globally from npm. Since the global typescript tsserver.js wasn't used for checking .ts files, it would give me the same error:

> Cannot compile modules unless the '--module' flag is provided

 but running tsc on the command-line would not. This is because the tsserver.js from TypeScript-Sublime-Plugin doesn't recognize some options in the tsconfig.json, like emitDecoratorMetadata for example.

TL;DR
The solution is to set the 'typescript_proc_file' settings in your sublime-settings file (Preferences -> Settings - User) to the globally installed tsserver.js file.
In my case:
`"typescript_proc_file": "C:/Users/admin/AppData/Roaming/npm/node_modules/typescript/lib/tsserver.js"`
You have to restart sublime for the changes to take effect.

**Comment by billti** [comment]
@SummerWish That behavior sounds correct. If you are listing the files explicitly in the tsconfig.json, and the current file isn't one of them (or included/imported by one of them), then the current files isn't going to use that tsconfig.json. It's going to get its own context.

@NeWbY100 Correct. The Sublime plugin ships its own version of tsserver.js, so if you want to use a specific version (i.e. one which may have features or behavior you depend on), then that needs to be specified.

With these two issues, and the prior issue of an invalid compilerOption, that seems to be all the root causes of this issue. I'll close now, comment for further investigation if there are further causes not covered. Thanks!

**Comment by NekR** [comment]
So why this is closed? TS plugin doesn't see my `tsconfig.json` for a new project. Yes, I reloaded sublime already. 

It doesn't see `d.ts` files in `files` section. Also it reports that I have to use `es6` target for es6 features, but I already specified that.

**Comment by NekR** [comment]
What about manually specifying path to `tsconfig.json` for a projects (my `tsconfig.json` is in the root)?

**Comment by NekR** [comment]
Or some debug UI to find `tsconfig.json` or possible errors? Because `webpack` TS loader correctly finds and uses `tsconfig.json`.

**Comment by mhegazy** [comment]
@NekR, can you share a sample we can look at?

**Comment by garkbit** [comment]
Echoing what @NekR is saying, I've tried the things mentioned above. `tsc` is quite happy with my `tsconfig.json` file, I've added the `typescript_proc_file` setting, and still no usage of my `tsconfig.json` file. I'm at a loss even for what to provide for diagnostic purposes.

**Comment by garkbit** [comment]
OK, so I manually downgraded the plugin to v1.1.1, and it correctly picks up `tsconfig.json`. The issue was introduced with v1.1.2.

**Comment by garkbit** [comment]
All right, I've resolved this. It was apparently basically the same issue face by @NeWbY100, however, the property that needed to be set has subsequently changed. Whereas @NeWbY100 set the following:
```
"typescript_proc_file": "/path/to/typescript/lib/tsserver.js"
```
The property and usage have now changed to:
```
"typescript_tsdk": "/path/to/typescript/lib"
```
Note the absence of `tsserver.js`. It's now used to figure out which `tsserver.js` _and_ `tsc.js` to use. 

@NekR, I'd give this a shot and see if it works.

**Comment by NekR** [comment]
@garkbit thanks for info, I will try it today or so.
@mhegazy sorry for not providing more info, missed your comment.

**Comment by johndeighan** [comment]
I'm having a similar problem with a default installation of Sublime Text 3. I'm learning Angular 2. To set up a new project, I'm using angular-cli like this:
<pre><code>
ng new angular2_hello_world
</code></pre>
which sets up this directory structure:
<pre>
angular2_hello_world/
   src/
      app/
         app.component.ts
         ...other files
      index.html
      tsconfig.json
      ...other files
</pre>
That is, my tsconfig.json is in the src/ folder, but outside the app/ folder. My source file, app.component.ts in in the app/ folder. When I open the source file, I see this error:
<pre>
Experimental support for decorators is a feature that is subject to change in a future release.
Set the 'experimentalDecorators' option to remove this warning.
</pre>
But that option is set to true in the tsconfig.json file, so I have to assume that my tsconfig.json is not being used.<br>
If it matters, I was cd's to the enclosing angular2_hello_world/ folder when I entered the command 'subl .'.<br>
Since this is the default directory structure created by angular-cli and I'm using a clean install of Sublime Text 3 (Build 3126), it seems to me that a lot of folks will run into this problem in the future. The web site builds and runs just fine - I'm not using Sublime Text 3 to build and run - but I'd really like that warning to go away and anticipate problems with other settings in the future if I can't get the Sublime TypeScript plug-in to use my tsconfig.json file.

**Comment by fiznool** [comment]
@johndeighan I think this is related to https://github.com/Microsoft/TypeScript-Sublime-Plugin/issues/542

**Comment by Quasso** [comment]
+1 @johndeighan, I've been experiencing that exact problem recently (notably, sporadically)
